### PR TITLE
Relax loadflow parameters for igm quality check

### PR DESCRIPTION
**Pull request by etiennehomer**
Signed-off-by: HOMER Etienne <etiennehomer@gmail.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java [hunk 1]
```diff
@@ -1,35 +1,33 @@
 /**
-=======
-package org.gridsuite.casevalidation.server; /**
->>>>>>> master
  * Copyright (c) 2020, RTE (http://www.rte-france.com)
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java:2]: Remove lines 2, 3, 4

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        LOGGER.info("Loadflow validation for case {} with loadflow parameters : {}", network.getId(), params);
+        boolean isLoadFlowOk = isMainComponentConverging(result);
+        LOGGER.info("Loadflow status: {}", isLoadFlowOk ? "OK" : "KO");
+        LOGGER.info("Loadflow metrics: {}", result.getMetrics());
+        return new LoadFlowCaseValidationReport(result.getMetrics(), isLoadFlowOk, params);
+    }
+
+    boolean isMainComponentConverging(LoadFlowResult result) {
+        //Open LoadFlow and HADES 2 return the main synchronous component result as component 0;
+        return result.getComponentResults().get(0).getStatus() == LoadFlowResult.ComponentResult.Status.CONVERGED;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:38]: Maybe add a test if getComponentResults() is null, which can occurs

File: src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java [hunk 3]
```diff
@@ -60,42 +70,201 @@ public void test() throws Exception {
...
+        //Loadflow diverges with both default and relaxed parameters
+        network.getGenerator("GEN_1").setTargetV(10);
+
+        given(networkStoreService.getNetwork(testNetworkId, PreloadingStrategy.COLLECTION)).willReturn(network);
+        result = mvc.perform(put("/v1/networks/{networkUuid}/validate", testNetworkId))
                 .andExpect(status().isOk())
                 .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                 .andReturn();
-        assertTrue(result.getResponse().getContentAsString().contains("status\":\"CONVERGED\""));
+        assertTrue(result.getResponse().getContentAsString().contains("\"validationOk\":false"));
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java:118]: You can remove this line

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java (hunk 1 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java:2]: Yes that's what i did

File: src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java (hunk 3 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java:118]: Done.

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java (hunk 2 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:38]: You're right. And also we don't want to check if the main component converges when then the loadflow is KO. I fixed it

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import com.powsybl.network.store.client.NetworkStoreService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Service;
+
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+@ComponentScan(basePackageClasses = {NetworkStoreService.class})
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:22]: You should move this to the Application so you have to dot it just one time.

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationService.java [hunk 5]
```diff
@@ -39,12 +48,31 @@ private Network getNetwork(UUID networkUuid) {
...
     }
 
-    CaseValidationReport validate(UUID networkUuid, LoadFlowParameters parameters) {
+    CaseValidationReport validate(UUID networkUuid) {
         Network network = getNetwork(networkUuid);
-        LoadFlowParameters params = parameters != null ? parameters : new LoadFlowParameters();
+        List<LoadFlowCaseValidationReport> loadFlowReports = new ArrayList<>();
+
+        //Validation with default loadflow parameters
+        LoadFlowParameters params = new LoadFlowParameters()
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/CaseValidationService.java:56]: All this code related to load flow based validation, should be move in LoadFlowCaseValidationReport

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import com.powsybl.loadflow.LoadFlowParameters;
+
+import java.util.Map;
+
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+public class LoadFlowCaseValidationReport {
+
+    private Map<String, String> metrics;
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java:18]: This is useless data.  What is important now is LoadFlowResult.ComponentResult.

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+@Service
+class LoadFlowCaseValidationService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LoadFlowCaseValidationService.class);
+
+    LoadFlowCaseValidationReport validate(Network network, LoadFlowParameters params) {
+        LoadFlowResult result = LoadFlow.run(network, params);
+        LOGGER.info("Loadflow validation for case {} with loadflow parameters : {}", network.getId(), params);
+        boolean isLoadFlowOk = result.isOk() && isMainComponentConverging(result);
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:30]: result.isOk() is useless and redundant with the other check.

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        LOGGER.info("Loadflow status: {}", isLoadFlowOk ? "OK" : "KO");
+        LOGGER.info("Loadflow metrics: {}", result.getMetrics());
+        return new LoadFlowCaseValidationReport(result.getMetrics(), isLoadFlowOk, params);
+    }
+
+    boolean isMainComponentConverging(LoadFlowResult result) {
+        //Open LoadFlow and HADES 2 return the main synchronous component result as component 0;
+        if (result.getComponentResults() == null || result.getComponentResults().isEmpty()) {
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:38]: result.getComponentResults() == null cannot append

Comment (geofjamg) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:38]: result.getComponentResults().isEmpty() won't append anymore in next OLF release. Add a comment to remove it later.

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+@ComponentScan(basePackageClasses = {NetworkStoreService.class})
+@Service
+class LoadFlowCaseValidationService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LoadFlowCaseValidationService.class);
+
+    LoadFlowCaseValidationReport validate(Network network, LoadFlowParameters params) {
+        LoadFlowResult result = LoadFlow.run(network, params);
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:28]: I think we should rely on loadflow server service and not directly use the LoadFlow java API, so that when we want to update the loadflow we ony have to do it in one micro service. I don't known why we did it like this...

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java [hunk 10]
```diff
@@ -1,35 +1,33 @@
...
-import java.util.Map;
+import java.util.List;
 
 /**
  * @author Etienne Homer <etienne.homer at rte-france.com>
  */
 public class CaseValidationReport {
 
-    private Map<String, String> metrics;
+    List<LoadFlowCaseValidationReport> loadFlowReports;
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java:16]: private is missing

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java (hunk 8 above)
Comment (geofjamg) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:38]: > result.getComponentResults() == null cannot append

From LoadFlowResultImpl.java
```java

    public LoadFlowResultImpl(boolean ok, Map<String, String> metrics, String logs, List<ComponentResult> componentResults) {
        this.ok = ok;
        this.metrics = Objects.requireNonNull(metrics);
        this.logs = logs;
        this.componentResults = Objects.requireNonNull(componentResults);
    }
```

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java (hunk 10 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/CaseValidationReport.java]: done

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java (hunk 7 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java]: done

File: src/main/java/org/gridsuite/casevalidation/server/CaseValidationService.java (hunk 5 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/CaseValidationService.java]: Done

**Review by FranckLecuyer — APPROVED** [review]

File: src/main/java/com/powsybl/loadflow/json/LoadFlowResultJsonModule.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ */
+package com.powsybl.loadflow.json;
+
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.powsybl.loadflow.LoadFlowResult;
+
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+public class LoadFlowResultJsonModule extends SimpleModule {
```
Comment (geofjamg) [inline src/main/java/com/powsybl/loadflow/json/LoadFlowResultJsonModule.java:15]: Already exists in core?

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package org.gridsuite.casevalidation.server;
+
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+public class LoadFlowCaseValidationReport {
+
+    private boolean ok;
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java:14]: Why do you keep this? This is redundant with status.

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        boolean isLoadFlowOk = validate(networkUuid, params);
+        return new LoadFlowCaseValidationReport(isLoadFlowOk, isLoadFlowOk ? LoadFlowCaseValidationReport.Status.CONVERGED_ON_2D_LF : LoadFlowCaseValidationReport.Status.FAILED);
+    }
+
+    boolean validate(UUID networkUuid, LoadFlowParameters params) {
+        LoadFlowResult result = loadFlowService.run(networkUuid, params);
+        LOGGER.info("Loadflow validation for case {} with loadflow parameters : {}", networkUuid, params);
+        boolean isLoadFlowOk = isMainComponentConverging(result);
+        LOGGER.info("Loadflow status: {}", isLoadFlowOk ? "OK" : "KO");
+        LOGGER.info("Loadflow metrics: {}", result.getMetrics());
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java:54]: Don't write metrics this is a useless and impl specific infos

File: src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java [hunk 14]
```diff
@@ -48,54 +49,68 @@
...
-        params.addExtension(OpenLoadFlowParameters.class, parametersExt);
-
-        ByteArrayOutputStream stream = new ByteArrayOutputStream();
-        JsonLoadFlowParameters.write(params, stream);
-        String paramsString = new String(stream.toByteArray());
-
-        result = mvc.perform(put("/v1/networks/{networkUuid}/validate", testNetworkId)
-                .contentType(MediaType.APPLICATION_JSON)
-                .content(paramsString))
+        JsonNode validationOk = om.readTree(result.getResponse().getContentAsString()).path("validationOk");
```
Comment (geofjamg) [inline src/test/java/org/gridsuite/casevalidation/server/CaseValidationTest.java:83]: You don't need t use an object mapper, you can directly assert with a andExpect 
https://mkyong.com/spring-boot/spring-test-how-to-test-a-json-array-in-jsonpath/

File: src/test/java/org/gridsuite/casevalidation/server/LoadFlowServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class LoadFlowServiceTest {
+
+    @Mock
+    private RestTemplate loadFlowServerRest;
+
+    private LoadFlowService loadFlowService;
+
+    private UUID randomUuid1 = UUID.randomUUID();
```
Comment (geofjamg) [inline src/test/java/org/gridsuite/casevalidation/server/LoadFlowServiceTest.java:42]: You should use a constant UUID, it is often better to always run unit tests with same data.

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java (hunk 12 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationReport.java]: done

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java (hunk 13 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java]: done

File: src/test/java/org/gridsuite/casevalidation/server/LoadFlowServiceTest.java (hunk 15 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/casevalidation/server/LoadFlowServiceTest.java]: done

File: src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java (hunk 9 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/casevalidation/server/LoadFlowCaseValidationService.java]: Done

**Review by geofjamg — APPROVED** [review]
